### PR TITLE
DAOS-6578 test: Refactor daos test python yaml files

### PR DESCRIPTION
**Pull request by sylviachanoiyee**
Refactor daos test python yaml files so we have granularity
to easily disable a subtest using skipForTicket.

### CODE REVIEW

File: src/tests/ftest/daos_test/daos_core_test-nvme_recovery.yaml [hunk 1]
```diff
@@ -47,20 +47,3 @@ dmg:
...
-      test_name: nvme_recovery tests 2
-      args: -u subtests="2"
-    test_n_3:
-      daos_test: N
-      test_name: nvme_recovery tests 3
-      args: -u subtests="3"
-    test_n_4:
-      daos_test: N
-      test_name: nvme_recovery tests 4
-      args: -u subtests="4"
```
Comment (phender) [inline src/tests/ftest/daos_test/daos_core_test-nvme_recovery.yaml:66]: Restore this without `!mux` and update the `test_*:` labels to match the `test_*` test method names.

File: src/tests/ftest/util/daos_core_base.py [hunk 2]
```diff
@@ -40,25 +40,15 @@ class DaosCoreBase(TestWithServers):
...
     def __init__(self, *args, **kwargs):
         """Initialize the DaosCoreBase object."""
         super(DaosCoreBase, self).__init__(*args, **kwargs)
-        self.subtest_name = None
 
-        test_timeout = self.params.get("test_timeout", self.TEST_PATH)
-        if test_timeout:
-            self.timeout = test_timeout
+        self.num_clients = None
+        self.scalable_ep = None
```
Comment (phender) [inline src/tests/ftest/util/daos_core_base.py:48]: As we discussed all of the changes to this file should be reverted and then this definition should be added before `test_timeout = self.params.get("test_timeout", self.TEST_PATH)` is called:
```
        self.TEST_PATH = "/run/daos_tests/Tests/{}".format(self.get_test_name())
```

**Review by phender — CHANGES_REQUESTED** [review]

File: src/tests/ftest/daos_test/daos_core_test-nvme_recovery.py [hunk 3]
```diff
@@ -27,19 +27,51 @@
 class DaosCoreTestNvme(DaosCoreBase):
     # pylint: disable=too-many-ancestors
     """Run just the daos_test NVMe Recovery tests.
+       Jira ID: DAOS-3846.
+
+       Test Description:
+           Purpose of this test is to run just the daos_test NVMe recovery
+           tests.
```
[the remark this answers is not in the document]
Comment (sylviachanoiyee) [reply src/tests/ftest/daos_test/daos_core_test-nvme_recovery.py:34]: ok

File: src/tests/ftest/daos_test/daos_core_test-nvme_recovery.py [hunk 4]
```diff
@@ -27,19 +27,51 @@
...
 
-    def test_nvme(self):
-        """Jira ID: DAOS-3846.
+    def test_nvme_1(self):
+        """Run daos_test -N subtest 1
+
+        :avocado: tags=all,hw,medium,nvme,ib2,full_regression
+        :avocado: tags=unittest,daos_test_nvme_recovery
+
+        """
```
[the remark this answers is not in the document]
Comment (sylviachanoiyee) [reply src/tests/ftest/daos_test/daos_core_test-nvme_recovery.py:45]: ok

File: src/tests/ftest/daos_test/daos_core_test-nvme_recovery.py [hunk 5]
```diff
@@ -27,19 +27,51 @@
...
+        DaosCoreBase.run_subtest(self, "N", "DAOS Nvme Recovery 1", 60,
+                                 "-u subtests='1'")
```
[the remark this answers is not in the document]
Comment (sylviachanoiyee) [reply src/tests/ftest/daos_test/daos_core_test-nvme_recovery.py:47]: ok

File: src/tests/ftest/daos_test/daos_core_test-nvme_recovery.yaml (hunk 1 above)
Comment (sylviachanoiyee) [reply src/tests/ftest/daos_test/daos_core_test-nvme_recovery.yaml:66]: ok

**Review by phender — COMMENTED** [review]
The code changes look good, but they do result in extremely long variant names.  Take daos_core_test.py for example (https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-4367/8/artifact/Functional/job-2021-01-25T18.54-67c9c19-daos_test-daos_core_test/results.html), one variant name is

`transport_config-test_daos_aggregate_ec-test_daos_array-test_daos_capability-test_daos_checksum-test_daos_container-test_daos_dedup-test_daos_degraded_ec_0to6-test_daos_degraded_ec_8to22-test_daos_degraded_mode-test_daos_distributed_tx-test_daos_drain_simple-test_daos_epoch-test_daos_epoch_recovery-test_daos_file_system-test_daos_io-test_daos_kv-test_daos_management-test_daos_md_replication-test_daos_object_array-test_daos_oid_allocator-test_daos_pool-test_daos_rebuild_ec-test_daos_rebuild_simple-test_daos_single_rdg_tx-test_daos_verify_consistency-num_clients-transport_config-hosts-pool-0-1-transport_config-efb8`

This is due to the layout of the test yaml params.  The parent path of each parameter is used to build the unique variant name.  From the following section of the daos_core_test.yaml file we get the "num_clients-test_daos_degraded_mode-test_daos_management-test_daos_pool-test_daos_container-..." part of the variant name:
```
daos_tests:
  num_clients:
    num_clients: 1
  Tests:
    test_daos_degraded_mode:
      daos_test: d
      test_name: DAOS_Degraded_Mode
      test_timeout: 450
      scalable_endpoint: True
    test_daos_management:
      daos_test: m
      test_name: DAOS_Management
      test_timeout: 110
    test_daos_pool:
      daos_test: p
      test_name: DAOS_Pool
      test_timeout: 120
    test_daos_container:
      daos_test: c
      test_name: DAOS_Container
      test_timeout: 135
    ...
```

We could get much shorter variant names with a slight reordering of the parameter paths such that the unique part (the test method name) is at the end:
```
daos_tests:
  num_clients:
    num_clients: 1
  daos_test:
    test_daos_degraded_mode: d
    test_daos_management: m
    test_daos_pool: p
    test_daos_container: c
    ...
  test_name:
    test_daos_degraded_mode: DAOS_Degraded_Mode
    test_daos_management: DAOS_Management
    test_daos_pool: DAOS_Pool
    test_daos_container: DAOS_Container
    ....
  test_timeout:
    test_daos_degraded_mode: 450
    test_daos_management: 110
    test_daos_pool: 120
    test_daos_container: 135
    ....
  scalable_endpoint:
    test_daos_degraded_mode: True
    ...
```

**Review by phender — APPROVED** [review]
It would be nice to have shorter variant names, but everything else looks good.

File: src/tests/ftest/daos_test/daos_core_test_dfs.py [hunk 6]
```diff
@@ -36,16 +35,36 @@ def __init__(self, *args, **kwargs):
         """Initialize the DaosCoreBase object."""
         super(DaosCoreTestDfs, self).__init__(*args, **kwargs)
         self.hostfile_clients_slots = None
+        self.daos_test = os.path.join(self.bin, 'dfs_test')
```
Comment (phender) [inline src/tests/ftest/daos_test/daos_core_test_dfs.py:38]: This assignment must be made in `setUp()` (after calling `super(<class>, self).setUp()`) or in a `test_*()` method as `self.bin` is not assigned until https://github.com/daos-stack/daos/blob/master/src/tests/ftest/util/apricot/apricot/test.py#L318.  As a result we hit this error:
```
[2021-01-27T21:19:46.228Z] Running: avocado run --ignore-missing-references on --html-job-result on --tap-job-result off --filter-by-tags=pr,hw,large --mux-yaml /tmp/tmpjs3m8r/daos_test-daos_core_test_dfs.yaml -- ./daos_test/daos_core_test_dfs.py
[2021-01-27T21:19:46.486Z] JOB ID     : 9346e8e418eb973cbf87f8816f29113fa6250ac6
[2021-01-27T21:19:46.486Z] JOB LOG    : /var/tmp/ftest/avocado/job-results/job-2021-01-27T21.19-9346e8e/job.log
[2021-01-27T21:19:47.072Z] Traceback (most recent call last):
[2021-01-27T21:19:47.072Z] 
[2021-01-27T21:19:47.072Z]   File "/usr/lib/python2.7/site-packages/avocado/core/job.py", line 453, in run_tests
[2021-01-27T21:19:47.072Z]     execution_order)
[2021-01-27T21:19:47.072Z] 
[2021-01-27T21:19:47.072Z]   File "/usr/lib/python2.7/site-packages/avocado/core/runner.py", line 599, in run_suite
[2021-01-27T21:19:47.072Z]     deadline):
[2021-01-27T21:19:47.072Z] 
[2021-01-27T21:19:47.072Z]   File "/usr/lib/python2.7/site-packages/avocado/core/runner.py", line 393, in run_test
[2021-01-27T21:19:47.072Z]     test_status.wait_for_early_status(proc, 60)
[2021-01-27T21:19:47.072Z] 
[2021-01-27T21:19:47.072Z] 
[2021-01-27T21:19:47.072Z] Avocado crashed: TestError: Process died before it pushed early test_status.
```
Due to:
```
2021-01-27 21:19:46,801 output           L0655 DEBUG| Traceback (most recent call last):
2021-01-27 21:19:46,802 output           L0655 DEBUG|   File "/usr/lib64/python2.7/multiprocessing/process.py", line 258, in _bootstrap
2021-01-27 21:19:46,802 output           L0655 DEBUG|     self.run()
2021-01-27 21:19:46,802 output           L0655 DEBUG|   File "/usr/lib64/python2.7/multiprocessing/process.py", line 114, in run
2021-01-27 21:19:46,802 output           L0655 DEBUG|     self._target(*self._args, **self._kwargs)
2021-01-27 21:19:46,802 output           L0655 DEBUG|   File "/usr/lib/python2.7/site-packages/avocado/core/runner.py", line 326, in _run_test
2021-01-27 21:19:46,802 output           L0655 DEBUG|     instance = loader.load_test(test_factory)
2021-01-27 21:19:46,802 output           L0655 DEBUG|   File "/usr/lib/python2.7/site-packages/avocado/core/loader.py", line 338, in load_test
2021-01-27 21:19:46,813 output           L0655 DEBUG|     test_instance = test_class(**test_parameters)
2021-01-27 21:19:46,813 output           L0655 DEBUG|   File "/usr/lib/daos/TESTING/ftest/daos_test/daos_core_test_dfs.py", line 38, in __init__
2021-01-27 21:19:46,814 output           L0655 DEBUG|     self.daos_test = os.path.join(self.bin, 'dfs_test')
2021-01-27 21:19:46,814 output           L0655 DEBUG|   File "/usr/lib64/python2.7/posixpath.py", line 77, in join
2021-01-27 21:19:46,814 output           L0655 DEBUG|     elif path == '' or path.endswith('/'):
2021-01-27 21:19:46,814 output           L0655 DEBUG| AttributeError: 'NoneType' object has no attribute 'endswith'
```

**Review by phender — CHANGES_REQUESTED** [review]

File: src/tests/ftest/daos_test/daos_core_test_dfs.py (hunk 6 above)
Comment (sylviachanoiyee) [reply src/tests/ftest/daos_test/daos_core_test_dfs.py:38]: thanks

File: src/tests/ftest/daos_test/daos_core_test.yaml [hunk 7]
```diff
@@ -57,104 +57,104 @@ dmg:
 daos_tests:
   num_clients:
     num_clients: 1
-  Tests: !mux
-    test_d:
+  Tests:
```
Comment (jolivier23) [inline src/tests/ftest/daos_test/daos_core_test.yaml:60]: what does the !mux do ?

**Review by jolivier23 — APPROVED** [review]

**Review by phender — APPROVED** [review]

**Pull request merged**